### PR TITLE
feat: add dialog, dialog-wrapper and underlay patterns

### LINKED ISSUE #565: Add Spectrum Dialog Pattern
### Expected Behaviour
https://opensource.adobe.com/spectrum-css/components/dialog/ should be available in the library

### Actual Behaviour
missing

#### The future
image

### PR DESCRIPTION
**Pull request by Westbrook**
Blocked by #655 

## Description
- add `<sp-dialog>`
- add `<sp-dialog-wrapper>`
- add `<sp-underlay>`
- demonstrate, document, test, repeat

## Related Issue
fixes #565 

## Motivation and Context
Coverage of Spectrum pattern.

## How Has This Been Tested?
- unit tests
- visual regressions

## Screenshots (if appropriate):
image
image
image

## Types of changes
- [x] New feature (non-breaking change which adds functionality)

## Checklist:
- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: packages/dialog/dialog-wrapper.md [hunk 1]
```diff
@@ -0,0 +1,86 @@
+## Description
+
+`sp-dialog` display important information that users need to acknowledge. They appear over the interface and block further interactions.
```
Comment (andrewhatran) [inline packages/dialog/dialog-wrapper.md:3]: I think you may have forgotten to change this to `sp-dialog-wrapper`

File: packages/dialog/dialog-wrapper.md [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+### Installation
+
+[![See it on NPM!](https://img.shields.io/npm/v/@spectrum-web-components/dialog?style=for-the-badge)](https://www.npmjs.com/package/@spectrum-web-components/dialog)
+[![How big is this package in your project?](https://img.shields.io/bundlephobia/minzip/@spectrum-web-components/dialog?style=for-the-badge)](https://bundlephobia.com/result?p=@spectrum-web-components/dialog)
+
+```
+npm install @spectrum-web-components/dialog
```
Comment (andrewhatran) [inline packages/dialog/dialog-wrapper.md:11]: ^ Same as above

File: packages/dialog/src/dialog-wrapper.ts [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+    public noDivider = false;
+
+    @property({ type: Boolean, reflect: true })
+    public open = false;
+
+    @property({ type: String, reflect: true })
+    public mode: 'fullscreen' | 'fullscreenTakeover' | '' = '';
+
+    @property({ type: String, reflect: true })
+    public size: 'small' | 'medium' | 'large' | 'alert' | '' = '';
```
Comment (andrewhatran) [inline packages/dialog/src/dialog-wrapper.ts:67]: Setting the default size to 'blank' is a but confusing for me. Is that equivalent to 'medium'?

File: packages/dialog/src/dialog.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    public noDivider = false;
+
+    @property({ type: Boolean, reflect: true })
+    public open = false;
+
+    @property({ type: String, reflect: true })
+    public mode: 'fullscreen' | 'fullscreenTakeover' | '' = '';
+
+    @property({ type: String, reflect: true })
+    public size: 'small' | 'medium' | 'large' | 'alert' | '' = '';
```
Comment (andrewhatran) [inline packages/dialog/src/dialog.ts:58]: Same as above

**Comment by andrewhatran** [comment]
Tested out the storybook and the component looks great! Nice job with the variation coverage and storybook examples.

Just a few comments about the README and size property, but other than that, it looks good to go 👍

File: packages/dialog/dialog-wrapper.md (hunk 2 above)
Comment (Westbrook) [reply packages/dialog/dialog-wrapper.md:11]: Here there is only the `@spectrum-web-components/dialog`, you can choose to use one component or the other, but both are included.

File: packages/dialog/dialog-wrapper.md (hunk 1 above)
Comment (Westbrook) [reply packages/dialog/dialog-wrapper.md:3]: Good catch!

File: packages/dialog/src/dialog-wrapper.ts (hunk 3 above)
Comment (Westbrook) [reply packages/dialog/src/dialog-wrapper.ts:67]: DId I write that? Hrm 🤔 It's the equivalent of it not being there...let me look deeper at what that does, and how it interacts with the styles. In some cases I've used a more complex getter/setter pair to ensure that the attribute was removed in this case, would that be more or less confusing?

Comment (Westbrook) [reply packages/dialog/src/dialog-wrapper.ts:67]: As we're not using this internally (beyond presence testing), would it be less confusing for this to be types as follows?
```
public size?: 'small' | 'medium' | 'large' | 'alert';
```

Comment (andrewhatran) [reply packages/dialog/src/dialog-wrapper.ts:67]: The above looks more clear to me, thanks! 

So a size of null is equivalent to the dialog not being active, that makes sense to me 👍

Comment (Westbrook) [reply packages/dialog/src/dialog-wrapper.ts:67]: `<sp-dialog open size="small|medium|large|alert">` enforces specific widths onto the dialog.

`<sp-dialog open>` will be open and take up as much space as it needs.

`<sp-dialog>` will not be active.

Comment (Westbrook) [reply packages/dialog/src/dialog-wrapper.ts]: Catch me up on how this feels now that that most recent code is up and passing CI.

Comment (andrewhatran) [reply packages/dialog/src/dialog-wrapper.ts]: This API looks solid to me 👍

**Pull request merged**